### PR TITLE
fix: ListUsers returns only the first error seen

### PR DESCRIPTION
**Pull request by miparnisari**
## Description
Proposal to fix flaky ListUsers tests, e.g.

```
--- FAIL: TestListUsersMemory/RunAll (0.00s)
          --- FAIL: TestListUsersMemory/RunAll/ListUsers (0.66s)
              --- FAIL: TestListUsersMemory/RunAll/ListUsers/relation_through_ttu_with_condition (0.21s)
                  --- FAIL: TestListUsersMemory/RunAll/ListUsers/relation_through_ttu_with_condition/stage_0 (0.20s)
                      --- FAIL: TestListUsersMemory/RunAll/ListUsers/relation_through_ttu_with_condition/stage_0/assertion_3 (0.01s)
                          listusers.go:187: 
                              	Error Trace:	/home/runner/work/openfga/openfga/tests/listusers/listusers.go:187
                              	Error:      	An error is expected but got nil.
                              	Test:       	TestListUsersMemory/RunAll/ListUsers/relation_through_ttu_with_condition/stage_0/assertion_3
                              	Messages:   	ListUsers request: object=document:1, relation=viewer, filters=user. Model: model
                              	            	  schema 1.1
                              	            	type user
                              	            	
                              	            	type folder
                              	            	  relations
                              	            	    define viewer: [user]
                              	            	
                              	            	type document
                              	            	  relations
                              	            	    define parent: [folder with str_cond, folder with xcond]
                              	            	    define viewer: [user] or viewer from parent
                              	            	
                              	            	condition str_cond(s: string) {
                              	            	  s == "hello"
                              	            	}
                              	            	
                              	            	condition xcond(x: int) {
                              	            	  x == 10
                              	            	}
                              	            	. Tuples: [user:"folder:a" relation:"parent" object:"document:1" condition:{name:"str_cond"} user:"folder:b" relation:"parent" object:"document:1" condition:{name:"xcond"} user:"user:jon" relation:"viewer" object:"folder:a" user:"user:jon" relation:"viewer" object:"folder:b"]. Contextual tuples: []
```

The root cause is that we have a `select` that is listening on two different channels, AND we're sending messages to the channel at roughly the same millisecond, which means [it's non-deterministic which message the `select` will see first](https://go.dev/ref/spec#Select_statements:~:text=If%20one%20or,communications%20can%20proceed).

Demo: https://go.dev/play/p/EixeM07Zolf, if you run enough times you will see the message `got here first`.

The proposal is to wait on the "done" channel first. After that, we either will have an error in the error channel, or we won't (but we won't have to block waiting for it).

## Related
- Original discussion https://github.com/openfga/openfga/pull/1714/files#r1656450644
- Supersedes https://github.com/openfga/openfga/pull/1719

### CODE REVIEW

File: pkg/server/commands/listusers/list_users_rpc.go [hunk 1]
```diff
@@ -216,26 +217,36 @@ func (l *listUsersQuery) ListUsers(
...
+		if deadlineExceeded || errors.Is(err, context.DeadlineExceeded) {
+			// We skip the error because we want to send at least partial results to the user (but we should probably set response headers)
+			break
+		}
+		telemetry.TraceError(span, err)
+		return nil, err
+	default:
+		break
+	}
+
```
Comment (willvedd) [inline pkg/server/commands/listusers/list_users_rpc.go:249]: This second `select`  statement is blocked either by a value on the `doneWithFoundUsersCh` or context cancellation. If an error gets emitted, I'm not sure that either that either will get triggered. This means that even if an error gets emitted, we're blocking until the context cancels. Or so that's my understanding currently. 

Can you demonstrate with a test that we fail as soon as the error gets emitted; that we're not always waiting for context to cancel?

Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc.go:249]: > If an error gets emitted, I'm not sure that either that either will get triggered.

The line `case <-doneWithFoundUsersCh:` will be executed after `foundUsersCh` is closed, which is after `expandErrCh <- resp.err`

Comment (willvedd) [reply pkg/server/commands/listusers/list_users_rpc.go:249]: >The line case <-doneWithFoundUsersCh: will be executed after foundUsersCh is closed, which is after expandErrCh <- resp.err.

Ok I see this now, makes sense.

I'd also like to double-check (ideally with a test) that the appropriate tests pass when setting a deadline to `0`.

File: pkg/server/list_users_test.go [hunk 2]
```diff
@@ -426,7 +427,7 @@ func TestListUsers_Deadline(t *testing.T) {
 
 		mockDatastore.EXPECT().
 			Read(gomock.Any(), storeID, gomock.Any()).
-			Return(nil, context.DeadlineExceeded).
```
Comment (willvedd) [inline pkg/server/list_users_test.go:429]: Why the change here?

File: internal/utils/concurrency.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package utils
```
Comment (willvedd) [inline internal/utils/concurrency.go:1]: **Non-blocking idea:** Would you consider having this concurrency package nested inside `internal` instead of being in the generic `utils` directory? It might make for a nicer import and implementation:

```
pool := concurrency.NewPool(ctx, int(l.resolveNodeBreadthLimit))
```

Or possibly:

```
concurrencyPool := pool.New(ctx, int(l.resolveNodeBreadthLimit))
```

**Review by willvedd — COMMENTED** [review]
I approve of this in principle but would like to see a test that ensures that the emitting of an error isn't blocked by context cancellation or the end of results. This doesn't appear to be the case but having confidence here would be great, especially since this is a such critical junction in the code.

File: internal/utils/concurrency.go (hunk 3 above)
Comment (miparnisari) [reply internal/utils/concurrency.go]: Moved it to `internal/concurrency`

**Comment by miparnisari** [comment]
> a test that ensures that the emitting of an error isn't blocked by context cancellation or the end of results

Sorry, I don't understand. I haven't changed how we emit errors, only how we consume them. When we emit an error, it will trigger the `case <-doneWithFoundUsersCh:` line

File: pkg/server/commands/listusers/list_users_rpc.go (hunk 1 above)
Comment (miparnisari) [reply pkg/server/commands/listusers/list_users_rpc.go:249]: Added `infinite_deadline_does_not_block_return_of_errors`

**Review by willvedd — APPROVED** [review]
